Argo/Globals/Parser.swift

```diff
@@ -0,0 +1,29 @@
+public enum Parser<T> {
+  case Success(Box<T>)
+  case TypeMismatch(String)
+  case MissingKey(String)
+
+  public var value: T? {
+    switch self {
+    case let .Success(box): return box.value
+    default: return .None
+    }
+  }
+}
+
+public func typeMismatch<T>(type: String, object: JSON) -> Parser<T> {
+  return .TypeMismatch("\(object) is not a \(type)")
+}
+
+public func parseOptional<A>(p: Parser<A>) -> Parser<A?> {
+  switch p {
+  case let .Success(box): return .Success(Box(.Some(box.value)))
+  case let .MissingKey(string): return .Success(Box(.None))
+  case let .TypeMismatch(string): return .TypeMismatch(string)
+  }
+}
+
+public final class Box<T> {
```

Comment:
Oh, right. Gross. Well, we should at least rely on an existing implementation then.